A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,296 @@
+object. Validation and processing are further described in
+[Stateful session][stateful-session]. Note that
+[StatefulSessionPerRoute][stateful-session-per-route] *will* be supported.
+
+The filter is described in more detail in the section
+[`CookieBasedStatefulSessionFilter` Processing][filter-section]
+
+### Load Balancer Configuration Containing `override_host_status`
+
+Even with stateful session affinity enabled, gRPC will only send an RPC to a
+valid backend (host) as defined by
+[`common_lb_config.override_host_status`][or-host-status]. When the backend
+state is invalid, gRPC will use the configured load balancing policy to pick
+the backend.
+
+Note that gRPC currently ignores endpoints with health status other than
+`HEALTHY` or `UNKNOWN` and even `XdsClient` excludes such endpoints in its
+update to the watchers. We will need a separate design to include such
+endpoints if we determine that such behavior is needed for this feature to
+be useful to our users.
+
+As a result, for this design gRPC will only consider the `UNKNOWN` or
+`HEALTHY` states that are specified in
+[`common_lb_config.override_host_status`][or-host-status] and ignore all
+other values. We considered the alternative of NACKing (i.e. rejecting)
+the CDS update when unsupported values are present, however it was discarded
+because of the difficulty in using such configuration in mixed deployments.
+
+The `override_host_status` value is included in the config update sent to
+the `“cluster_resolver_experimental”` policy which passes down the update to
+the `“xds_cluster_impl”` policy where it is used to create the additional
+child policy as described below.
+
+### `CookieBasedStatefulSessionFilter` Processing
+
+A channel configured for stateful session affinity will have the
+`CookieBasedStatefulSessionFilter` installed (interceptor in Java/Go).
+This filter will process incoming RPCs and set an appropriate LB pick value
+which will be passed to the Load Balancing Policy's Pick method.
+
+When an RPC arrives on a configured channel the Config Selector processes it
+as described in [A31: gRPC xDS Config Selector Design][A31]. This includes the
+`CookieBasedStatefulSessionFilter` which is installed as one of the filters in
+the channel. Note that the filter maintains a context for an RPC and processes
+both the outgoing request and incoming response of the RPC within that context.
+Also note that the filter is initialized with 3 configuration values - `name`,
+`ttl`  and `path` as described above.
+
+The filter performs the following steps:
+
+* If the incoming RPC “path” does not match the `path` configuration value,
+skip any further processing and just pass the RPC through (to the next
+filter). For example if the `path` configuration value is `/Package1.Service2`
+and the RPC method is `/Package2.Service1/Method3` then just pass the RPC
+through.
+
+* Search the RPC headers (metadata) for header(s) named `Cookie` and get the
```

Comment:
There doesn't appear to be any way to validate that the client hasn't tampered with the cookie. Is this only intended to be used with "trusted" clients? Should that be called out?

Reply:
Correct - there is no way to validate the cookie in the current implementation. We can make the cookie tamperproof by using cryptography or adding state but that's not a requirement for now.

The worst a client can do by tampering with the cookie is to send the RPC to a specific backend (host) but it has to be a valid backend in the cluster otherwise gRPC will send the RPC through the normal LB policy. With a valid backend a malicious client can launch a DoS attack. But note that the Envoy implementation has the same limitation and the ring-hash feature also can be misused by malicious clients for the DoS attack.

Reply:
I think it would be useful for gRPC to provide a hook/interface to allow the user code some control over the cookie. While in a direct path ( gRPC Client -> gRPC server ) we can control the content, there are other cases - like client -> ingress LB -> server where the ingress LB may need to route.

It is worth looking at how Serverless and other similar services in google handle this. 

Also: for a gRPC client -> Envoy sidecar -> proxyless gRPC server, we need to make sure the handling of the cookie is consistent and works.

Reply:
> I think it would be useful for gRPC to provide a hook/interface to allow the user code some control over the cookie.

The standard cookie jar implementations may work? Or can you suggest an interface that will be useful?

> for a gRPC client -> Envoy sidecar -> proxyless gRPC server, we need to make sure the handling of the cookie is consistent and works.

Correct and it is consistent as per this gRFC.

Reply:
I think a standard cookie jar in an interceptor can do 99% of the work. I mentioned in a different comment that clients 
can already take the IP and directly Dial(IP) for subsequent requests - bypassing the service LB, but it's likely inefficient and   ugly. What would help IMO is some interface or setting to indicate to the LB to use the cookie - similar to what the xDS filter does, but exposed as an API/setting/envvar so client code can use it directly, without xDS.

Reply:
In gRPC, service config is the way to add this setting (instead of env var) and that's what Mark was suggesting. Let me think about it.

Reply:
I think we've already hashed out these issues in subsequent threads, so I'm resolving this comment.